br/pkg/storage/gcs_extra.go

```diff
@@ -0,0 +1,418 @@
+	workers     int
+	totalSize   int64
+	uploadID    string
+	chunkCh     chan chunk
+	curPart     int
+}
+
+// NewGCSWriter returns a GCSWriter which uses GCS multipart upload API behind the scene.
+func NewGCSWriter(
+	ctx context.Context,
+	cli *storage.Client,
+	uri string,
+	partSize int64,
+	parallelCnt int,
+	bucketName string,
+) (*GCSWriter, error) {
+	if partSize < gcsMinimumChunkSize || partSize > gcsMaximumChunkSize {
+		return nil, fmt.Errorf(
+			"invalid chunk size: %d. Chunk size must be between %d and %d",
+			partSize, gcsMinimumChunkSize, gcsMaximumChunkSize,
+		)
+	}
+
+	w := &GCSWriter{
+		uploadBase: uploadBase{
+			ctx:             ctx,
+			cli:             cli,
+			bucket:          bucketName,
+			blob:            uri,
+			retry:           defaultRetry,
+			signedURLExpiry: defaultSignedURLExpiry,
+		},
+		chunkSize: partSize,
+		workers:   parallelCnt,
+	}
+	if err := w.init(); err != nil {
+		return nil, fmt.Errorf("failed to initiate GCSWriter: %w", err)
+	}
+
+	return w, nil
+}
+
+func (w *GCSWriter) init() error {
+	opts := &storage.SignedURLOptions{
+		Scheme:          storage.SigningSchemeV4,
+		Method:          "POST",
+		Expires:         time.Now().Add(w.signedURLExpiry),
+		QueryParameters: url.Values{mpuInitiateQuery: []string{""}},
+	}
+	u, err := w.cli.Bucket(w.bucket).SignedURL(w.blob, opts)
+	if err != nil {
+		return fmt.Errorf("Bucket(%q).SignedURL: %s", w.bucket, err)
+	}
+
+	client := resty.New()
+	resp, err := client.R().Post(u)
+	if err != nil {
+		return fmt.Errorf("POST request failed: %s", err)
+	}
+
+	if resp.StatusCode() != http.StatusOK {
+		return fmt.Errorf("POST request returned non-OK status: %d", resp.StatusCode())
+	}
+	body := resp.Body()
+
+	result := InitiateMultipartUploadResult{}
+	err = xml.Unmarshal(body, &result)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal response body: %s", err)
+	}
+
+	uploadID := result.UploadId
+	w.uploadID = uploadID
+	w.chunkCh = make(chan chunk, w.workers)
+	for i := 0; i < w.workers; i++ {
+		w.wg.Add(1)
+		go w.readChunk(w.chunkCh)
+	}
+	w.curPart = 1
+	return nil
+}
+
+func (w *GCSWriter) readChunk(ch chan chunk) {
+	defer w.wg.Done()
+	for {
+		data, ok := <-ch
+		if !ok {
+			break
+		}
+
+		select {
+		case <-w.ctx.Done():
+			data.cleanup()
+			return
+		default:
+			part := &xmlMPUPart{
+				uploadBase: w.uploadBase,
+				uploadID:   w.uploadID,
+				buf:        data.buf,
+				partNumber: data.num,
+			}
+			if w.err.Load() == nil {
+				if err := part.Upload(); err != nil {
+					w.err.Store(err)
+				}
+			}
+			w.appendMPUPart(part)
+			data.cleanup()
+		}
+	}
+}
+
+// Write uploads given bytes as a part to Google Cloud Storage. Write is not
+// concurrent safe.
+func (w *GCSWriter) Write(p []byte) (n int, err error) {
+	if w.curPart > gcsMaximumParts {
+		err = fmt.Errorf("exceed maximum parts %d", gcsMaximumParts)
+		if w.err.Load() == nil {
+			w.err.Store(err)
+		}
+		return 0, err
+	}
+	buf := make([]byte, len(p))
```

Comment:
use sync.Pool to cache the []byte will OOM 😂 I'll check if a chan []byte is better in next week